### PR TITLE
Introduce a simple state sort for drawing opaque items

### PR DESCRIPTION
**Pull request by samcake**
a Simple state sort applied to the Opaque Items in the new DrawStateSortDeferred Job used in place of the DrawDeferred job.

THis is sorting the items based on the ShapeKey which allow the code to factorise the setPipeline and greatly reduce the number of state changes.
THis also as a nice impact on performance on the batch rendering as the glUseProgram is used much more efficiently.
This is a very simple first step toward better state sort but this already gives use nice performance improvment.

in the DrawOpaqueDeferred job I measure ~1ms over 10ms in Playa in stereo.
The state Sort can be turned off so we can compare in real time the differrences of performance.

Added the counter "frameSetPipelineCounter" in Render.Stats that represent the number of time the SetPipeline is actually performed in the GLBackend layers with true glUseProgram beeing called.

Cleaned up the renderStats.js qml to introduce a new plot "State Changes" which shows this new SetPipeline counter and migrated the Frame Texture Counter  there.
Removed the per Frame Texture Memory since it was a bit confusing.

### CODE REVIEW

**Comment by samcake** [comment]
To Test:
Compare performances between Master and This PR build in any domain. This pr build should be faster.
Looking at the detailed timing stats, note the ms value for DrawOpaqueDeferred. THis should be better  (sorter) with this PR.

Using the PR, and the renderStats.js from this PR (the qml has changed) note the new PLot called "State Changes" representing 2 curves, one for Textures and one for Pipelines

Through the js console, access the field:
Render.RenderDeferredTask.DrawOpaqueDeferred.stateSort

It should exists and be true at start up
turn it false and note that the performances are going worse (normally back to master)
THe number of pipelines changes should increase when state sort is off, the number of textures bound too

File: libraries/render/src/render/DrawTask.cpp [hunk 1]
```diff
@@ -65,6 +65,57 @@ void render::renderShapes(const SceneContextPointer& sceneContext, const RenderC
...
+                bucket.push_back(item);
+            } else if (key.hasOwnPipeline()) {
+                ownPipelineBucket.push_back(item);
+            } else {
+                qDebug() << "Item could not be rendered: invalid key ?" << key;
+            }
+        }
+    }
+
+    // THen render
```
Comment (ZappoMan) [inline libraries/render/src/render/DrawTask.cpp]: TYpo

**Comment by hifi-gustavo** [comment]
Build console output for commit: d746fba142d271aed425c61a5b9ac98a82363fe9 available at: https://gist.github.com/6b520bd9a2f4b9901586b0e49ae360aa<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7697.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7697.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 0ae54b9abc9d894d01c6501460231f94db249c3b available at: https://gist.github.com/6b520bd9a2f4b9901586b0e49ae360aa<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7697.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7697.dmg

File: libraries/render-utils/src/RenderDeferredTask.cpp [hunk 2]
```diff
@@ -201,6 +201,38 @@ void DrawDeferred::run(const SceneContextPointer& sceneContext, const RenderCont
     });
 }
 
+void DrawStateSortDeferred::run(const SceneContextPointer& sceneContext, const RenderContextPointer& renderContext, const ItemBounds& inItems) {
```
Comment (zzmp) [inline libraries/render-utils/src/RenderDeferredTask.cpp]: Can we move the state sort into a new header? This file is very large.

File: libraries/render-utils/src/RenderDeferredTask.cpp [hunk 3]
```diff
@@ -201,6 +201,38 @@ void DrawDeferred::run(const SceneContextPointer& sceneContext, const RenderCont
...
+
+        glm::mat4 projMat;
+        Transform viewMat;
+        args->_viewFrustum->evalProjectionMatrix(projMat);
+        args->_viewFrustum->evalViewTransform(viewMat);
+
+        batch.setProjectionTransform(projMat);
+        batch.setViewTransform(viewMat);
+
+        if (_stateSort) {
```
Comment (zzmp) [inline libraries/render-utils/src/RenderDeferredTask.cpp]: Is this switch just for debugging? Does it belong in the codebase?

File: libraries/render-utils/src/RenderDeferredTask.h [hunk 4]
```diff
@@ -72,6 +72,44 @@ class DrawDeferred {
...
+};
+
+class DrawStateSortDeferred {
+public:
+    using Config = DrawStateSortConfig;
+    using JobModel = render::Job::ModelI<DrawStateSortDeferred, render::ItemBounds, Config>;
+
+    DrawStateSortDeferred(render::ShapePlumberPointer shapePlumber) : _shapePlumber{ shapePlumber } {}
+
+    void configure(const Config& config) { _maxDrawn = config.maxDrawn; _stateSort = config.stateSort; }
```
Comment (zzmp) [inline libraries/render-utils/src/RenderDeferredTask.h]: This implementation should be in the cpp.

File: libraries/render/src/render/DrawTask.cpp [hunk 5]
```diff
@@ -65,6 +65,57 @@ void render::renderShapes(const SceneContextPointer& sceneContext, const RenderC
     }
 }
 
+void render::renderStateSortShapes(const SceneContextPointer& sceneContext, const RenderContextPointer& renderContext,
+    const ShapePlumberPointer& shapeContext, const ItemBounds& inItems, int maxDrawnItems) {
+    auto& scene = sceneContext->_scene;
+    RenderArgs* args = renderContext->args;
+
+    int numItemsToDraw = (int)inItems.size();
```
Comment (zzmp) [inline libraries/render/src/render/DrawTask.cpp]: Eventually, we should clean up these repeated bits of code.

File: libraries/render/src/render/DrawTask.cpp [hunk 6]
```diff
@@ -65,6 +65,57 @@ void render::renderShapes(const SceneContextPointer& sceneContext, const RenderC
...
+    if (maxDrawnItems != -1) {
+        numItemsToDraw = glm::min(numItemsToDraw, maxDrawnItems);
+    }
+
+    using SortedPipelines = std::vector< render::ShapeKey >;
```
Comment (zzmp) [inline libraries/render/src/render/DrawTask.cpp]: There aren't spaces around templated types elsewhere.

File: libraries/render/src/render/DrawTask.cpp [hunk 7]
```diff
@@ -65,6 +65,57 @@ void render::renderShapes(const SceneContextPointer& sceneContext, const RenderC
...
+    using SortedShapes = std::unordered_map< render::ShapeKey, std::vector< Item >, render::ShapeKey::Hash, render::ShapeKey::KeyEqual >;
+    SortedPipelines sortedPipelines;
```
Comment (zzmp) [inline libraries/render/src/render/DrawTask.cpp]: Do you need this? Can't you look up the key by iterating over the unordered_map and checking the iterator key?

**Comment by samcake** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 0ae54b9abc9d894d01c6501460231f94db249c3b available at: https://gist.github.com/6b520bd9a2f4b9901586b0e49ae360aa<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7697.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7697.dmg

File: libraries/render-utils/src/RenderDeferredTask.cpp (hunk 3 above)
Comment (samcake) [inline libraries/render-utils/src/RenderDeferredTask.cpp]: Yes it s mostly for debugging
I wanted a way to prove the benfirt in realtime
I can remove that

Comment (samcake) [inline libraries/render-utils/src/RenderDeferredTask.cpp]: I m going to keep it there though so QA can test, i can remove it in a later PR

File: libraries/render-utils/src/RenderDeferredTask.h (hunk 4 above)
Comment (samcake) [inline libraries/render-utils/src/RenderDeferredTask.h]: it s ONE line!

File: libraries/render/src/render/DrawTask.cpp (hunk 5 above)
Comment (samcake) [inline libraries/render/src/render/DrawTask.cpp]: I Agree :)

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9087de22bd2a52d462ca0e22677233281385a5ef available at: https://gist.github.com/6b520bd9a2f4b9901586b0e49ae360aa<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7697.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7697.dmg

**Comment by OmegaHeron** [comment]
Tested Windows7 
DrawOpaqueDeferred ran around 2...5ms in Home scene with this PR and 3...6ms in stable.  In all locations tested with PR uses less time in drawOpaqueDeffered.

Tested toggle of Render.RenderDeferredTask.DrawOpaqueDeferred.stateSort and it reacted as test notes indicate it should.

Slightly better FPS at Cellscience "Inside the Cell" location.
One crash recorded via bugsplat at Playa (as always).

General smoke test passes.

**Pull request merged**